File: ypsopump/src/main/java/info/nightscout/androidaps/plugins/pump/ypsopump/util/YpsoPumpUtil.java
package info.nightscout.androidaps.plugins.pump.ypsopump.util;

import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.nio.ByteBuffer;
import java.security.InvalidParameterException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

import javax.inject.Inject;

import info.nightscout.androidaps.logging.AAPSLogger;
import info.nightscout.androidaps.logging.LTag;
import info.nightscout.androidaps.plugins.bus.RxBusWrapper;
import info.nightscout.androidaps.plugins.pump.common.defs.PumpDriverState;
import info.nightscout.androidaps.plugins.pump.common.utils.ByteUtil;
import info.nightscout.androidaps.plugins.pump.ypsopump.defs.YpsoPumpCommandType;
import info.nightscout.androidaps.plugins.pump.ypsopump.driver.YpsopumpPumpStatus;
import info.nightscout.androidaps.plugins.pump.ypsopump.event.EventPumpStatusChanged;

public class YpsoPumpUtil {

    public static final int MAX_RETRY = 2;
    private final AAPSLogger aapsLogger;
    private final RxBusWrapper rxBus;
    private final YpsopumpPumpStatus ypsopumpPumpStatus;

    private static PumpDriverState driverStatus = PumpDriverState.Sleeping;
    private YpsoPumpCommandType pumpCommandType;
    public Gson gson = new GsonBuilder().setPrettyPrinting().create();

    @Inject
    public YpsoPumpUtil(
            AAPSLogger aapsLogger,
            RxBusWrapper rxBus,
            YpsopumpPumpStatus ypsopumpPumpStatus
    ) {
        this.aapsLogger = aapsLogger;
        this.rxBus = rxBus;
        this.ypsopumpPumpStatus = ypsopumpPumpStatus;
    }

    public YpsopumpPumpStatus getYpsopumpPumpStatus() {
        return this.ypsopumpPumpStatus;
    }


    public PumpDriverState getDriverStatus() {
        return (PumpDriverState) workWithStatusAndCommand(StatusChange.GetStatus, null, null);
    }

    public YpsoPumpCommandType getCurrentCommand() {
        return (YpsoPumpCommandType) workWithStatusAndCommand(StatusChange.GetCommand, null, null);
    }

    public void resetDriverStatusToConnected() {
        workWithStatusAndCommand(StatusChange.SetStatus, PumpDriverState.Connected, null);
    }

    public void setDriverStatus(PumpDriverState status) {
        aapsLogger.debug(LTag.PUMP, "Set driver status: " + status.name());
        workWithStatusAndCommand(StatusChange.SetStatus, status, null);
    }

    public void setCurrentCommand(YpsoPumpCommandType currentCommand) {
        aapsLogger.debug(LTag.PUMP, "Set current command: " + currentCommand.name());
        workWithStatusAndCommand(StatusChange.SetCommand, PumpDriverState.ExecutingCommand, currentCommand);
    }

    public synchronized Object workWithStatusAndCommand(StatusChange type, PumpDriverState driverStatusIn, YpsoPumpCommandType pumpCommandType) {

        //aapsLogger.debug(LTag.PUMP, "Status change type: " + type.name() + ", DriverStatus: " + (driverStatus != null ? driverStatus.name() : ""));

        switch (type) {

            case GetStatus:
                //aapsLogger.debug(LTag.PUMP, "GetStatus: DriverStatus: " + driverStatus);
                return driverStatus;

            case GetCommand:
                return this.pumpCommandType;

            case SetStatus: {
                //aapsLogger.debug(LTag.PUMP, "SetStatus: DriverStatus: " + driverStatus + ", Incoming: " + driverStatusIn);
                driverStatus = driverStatusIn;
                this.pumpCommandType = null;
                //aapsLogger.debug(LTag.PUMP, "SetStatus: DriverStatus: " + driverStatus);
                rxBus.send(new EventPumpStatusChanged());
            }
            break;

            case SetCommand: {
                this.driverStatus = driverStatusIn;
                this.pumpCommandType = pumpCommandType;
                rxBus.send(new EventPumpStatusChanged());
            }
            break;
        }

        return null;
    }


    public void sleepSeconds(long seconds) {
        try {
            Thread.sleep(seconds * 1000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }


    public void sleep(long miliseconds) {
        try {
            Thread.sleep(miliseconds);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }


    public static boolean isSame(Double d1, Double d2) {
        double diff = d1 - d2;

        return (Math.abs(diff) <= 0.000001);
    }

    public static boolean isSame(Double d1, Integer d2) {
        double diff = d1 - d2;

        return (Math.abs(diff) <= 0.000001);
    }

    private enum StatusChange {
        GetStatus,
        GetCommand,
        SetStatus,
        SetCommand
    }


    public static byte[] computeUserLevelPassword(String btAddress) {
        byte[] sourceArray = new byte[]{91, (byte) 215, 16, 102, 1, (byte) 242, 79, 60, (byte) 143, 107};
        byte[] array = new byte[16];
        String[] address = btAddress.split(":");
        Log.d("DD", "BTAddress: " + address);
        array[0] = convertIntegerStringToByte(address[0]);
        array[1] = convertIntegerStringToByte(address[1]);
        array[2] = convertIntegerStringToByte(address[2]);
        array[3] = convertIntegerStringToByte(address[3]);
        array[4] = convertIntegerStringToByte(address[4]);
        array[5] = convertIntegerStringToByte(address[5]);
        System.arraycopy(sourceArray, 0, array, 6, 10);
//        return MD5.Create().ComputeHash(array);

//        MessageDigest.getInstance(MD5)
//
//        DigestUtils.md5Hex(password).toUpperCase();

        MessageDigest md = null;
        try {
            md = MessageDigest.getInstance("MD5");
            md.update(array);
            return md.digest();
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
            return array;
        }


        //return array;
    }

    private static byte convertIntegerStringToByte(String hex) {
        //Log.d("DD", "BTAddress Part: " + hex);
        return (new Integer(Integer.parseInt(hex, 16))).byteValue();
    }


    public static byte[] getSettingIdAsArray(int settingId) {
        byte[] array = new byte[8];
        CreateGLB_SAFE_VAR(settingId, array);
        return array;
    }

    public static byte[] getBytesFromInt16(int value) {
        byte[] array = getBytesFromInt(value);
        Log.d("HHH", ByteUtil.getHex(array));
        return new byte[]{array[3], array[2]}; // 2, 3
    }

    public static byte[] getBytesFromInt(int value) {
        return ByteBuffer.allocate(4).putInt(value).array();
    }

    //public static byte[]


    // TOOO fix
    public static void CreateGLB_SAFE_VAR(int piValue, byte[] pbyBuffer) {
        pbyBuffer[0] = (byte) piValue;
        piValue >>= 8;
        pbyBuffer[1] = (byte) piValue;
        piValue >>= 8;
        pbyBuffer[2] = (byte) piValue;
        piValue >>= 8;
        pbyBuffer[3] = (byte) piValue;
        pbyBuffer[4] = (byte) (~pbyBuffer[0]);
        pbyBuffer[5] = (byte) (~pbyBuffer[1]);
        pbyBuffer[6] = (byte) (~pbyBuffer[2]);
        pbyBuffer[7] = (byte) (~pbyBuffer[3]);
    }

    public static int getValueFromeGLB_SAFE_VAR(byte[] paBuffer) {
        if ((~paBuffer[7] & 0xFF) != (paBuffer[3] & 0xFF) || (~paBuffer[6] & 0xFF) != (paBuffer[2] & 0xFF) || (~paBuffer[5] & 0xFF) != (paBuffer[1] & 0xFF) || (~paBuffer[4] & 0xFF) != (paBuffer[0] & 0xFF)) {
            throw new InvalidParameterException("Invalid GLB_SAFE_VAR byte array:" + paBuffer);
        }
        return byteToInt(paBuffer, 0, 4);
    }


    public static int byteToInt(byte[] data, int start, int length) {
        if (length == 1) {
            return ByteUtil.toInt(data[start]);
        } else if (length == 2) {
            return ByteUtil.toInt(data[start], data[start + 1], ByteUtil.BitConversion.LITTLE_ENDIAN);
        } else if (length == 3) {
            return ByteUtil.toInt(data[start], data[start + 1], data[start + 2], ByteUtil.BitConversion.LITTLE_ENDIAN);
        } else if (length == 4) {
            return ByteUtil.toInt(data[start], data[start + 1], data[start + 2], data[start + 3], ByteUtil.BitConversion.LITTLE_ENDIAN);
        } else {
            //aapsLogger.error(LTag.PUMPBTCOMM, "byteToInt, length $length not supported.");
            return 0;
        }
    }


    public static byte[] getBytesFromIntArray2(int value) {
        byte[] array = ByteBuffer.allocate(4).putInt(value).array();
        return new byte[]{array[3], array[2]};
    }

}

File: ypsopump/src/main/java/info/nightscout/androidaps/plugins/pump/ypsopump/util/YpsoPumpUtil.kt
package info.nightscout.androidaps.plugins.pump.ypsopump.util

import android.util.Log
import com.google.gson.GsonBuilder
import info.nightscout.androidaps.plugins.bus.RxBus
import info.nightscout.androidaps.plugins.general.overview.events.EventNewNotification
import info.nightscout.androidaps.plugins.general.overview.notifications.Notification
import info.nightscout.androidaps.plugins.pump.common.defs.PumpDriverState
import info.nightscout.androidaps.plugins.pump.common.utils.ByteUtil
import info.nightscout.androidaps.plugins.pump.ypsopump.comm.ble.defs.YpsoPumpNotificationType
import info.nightscout.androidaps.plugins.pump.ypsopump.data.DateTimeDto
import info.nightscout.androidaps.plugins.pump.ypsopump.defs.YpsoPumpCommandType
import info.nightscout.androidaps.plugins.pump.ypsopump.defs.YpsoPumpErrorType
import info.nightscout.androidaps.plugins.pump.ypsopump.driver.YpsopumpPumpStatus
import info.nightscout.androidaps.plugins.pump.ypsopump.event.EventPumpStatusChanged
import info.nightscout.androidaps.utils.resources.ResourceHelper
import info.nightscout.shared.logging.AAPSLogger
import info.nightscout.shared.logging.LTag
import java.nio.ByteBuffer
import java.security.InvalidParameterException
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.experimental.and
import kotlin.experimental.inv

@Singleton
class YpsoPumpUtil @Inject constructor(
    val aapsLogger: AAPSLogger,
    val rxBus: RxBus,
    val resourceHelper: ResourceHelper,
    val ypsopumpPumpStatus: YpsopumpPumpStatus
) {

    var preventConnect: Boolean = false

    //private var driverStatusInternal: PumpDriverState
    private var pumpCommandType: YpsoPumpCommandType? = null
    var gson = GsonBuilder().setPrettyPrinting().create()



    // var driverStatus: PumpDriverState
    //     get() = workWithStatusAndCommand(StatusChange.GetStatus, null, null) as PumpDriverState?

    // val currentCommand: YpsoPumpCommandType?
    //     get() = workWithStatusAndCommand(StatusChange.GetCommand, null, null) as YpsoPumpCommandType?

    fun resetDriverStatusToConnected() {
        workWithStatusAndCommand(StatusChange.SetStatus, PumpDriverState.Ready, null)
    }

    var driverStatus: PumpDriverState
        get() {
            var stat = workWithStatusAndCommand(StatusChange.GetStatus, null, null) as PumpDriverState
            aapsLogger.debug(LTag.PUMP, "Get driver status: " + stat.name)
            return stat
        }
        set(status) {
            aapsLogger.debug(LTag.PUMP, "Set driver status: " + status.name)
            workWithStatusAndCommand(StatusChange.SetStatus, status, null)
        }

    var currentCommand: YpsoPumpCommandType
        get() = workWithStatusAndCommand(StatusChange.GetCommand, null, null) as YpsoPumpCommandType
        set(currentCommand) {
            aapsLogger.debug(LTag.PUMP, "Set current command: " + currentCommand.name)
            workWithStatusAndCommand(StatusChange.SetCommand, PumpDriverState.ExecutingCommand, currentCommand)
        }

    var errorType: YpsoPumpErrorType?
        get() {
            return workWithStatusAndCommand(StatusChange.GetError, null, null) as YpsoPumpErrorType?
        }
        set(error) {
            workWithStatusAndCommand(StatusChange.SetError, null, null, error)
        }

    // fun getDriverStatus(): PumpDriverState {
    //     return workWithStatusAndCommand(StatusChange.GetStatus, null, null) as PumpDriverState
    // }
    //
    // fun setDriverStatus(status: PumpDriverState) {
    //     aapsLogger.debug(LTag.PUMP, "Set driver status: " + status.name)
    //     workWithStatusAndCommand(StatusChange.SetStatus, status, null)
    // }

    // fun getCurrentCommand(): YpsoPumpCommandType {
    //     return workWithStatusAndCommand(StatusChange.GetCommand, null, null) as YpsoPumpCommandType
    // }

    // fun setCurrentCommand(currentCommand: YpsoPumpCommandType) {
    //     aapsLogger.debug(LTag.PUMP, "Set current command: " + currentCommand.name)
    //     workWithStatusAndCommand(StatusChange.SetCommand, PumpDriverState.ExecutingCommand, currentCommand)
    // }

    @Synchronized
    fun workWithStatusAndCommand(type: StatusChange,
                                 driverStatusIn: PumpDriverState?,
                                 pumpCommandType: YpsoPumpCommandType?,
                                 ypsoPumpErrorType: YpsoPumpErrorType? = null): Any? {

        //aapsLogger.debug(LTag.PUMP, "Status change type: " + type.name() + ", DriverStatus: " + (driverStatus != null ? driverStatus.name() : ""));
        when (type) {
            StatusChange.GetStatus  ->  {
                aapsLogger.debug(LTag.PUMP, "GetStatus: DriverStatus: " + driverStatusInternal);
                return driverStatusInternal
            }
            StatusChange.SetStatus  -> {
                aapsLogger.debug(LTag.PUMP, "SetStatus: DriverStatus Before: " + driverStatusInternal + ", Incoming: " + driverStatusIn);
                driverStatusInternal = driverStatusIn!!
                this.pumpCommandType = null
                aapsLogger.debug(LTag.PUMP, "SetStatus: DriverStatus: " + driverStatusInternal);
                rxBus.send(EventPumpStatusChanged(driverStatusInternal))
            }
            StatusChange.GetCommand -> return this.pumpCommandType
            StatusChange.SetCommand -> {
                driverStatusInternal = driverStatusIn!!
                this.pumpCommandType = pumpCommandType
                rxBus.send(EventPumpStatusChanged(driverStatusInternal))
            }
            StatusChange.GetError   -> return errorTypeInternal
            StatusChange.SetError   -> {
                errorTypeInternal = ypsoPumpErrorType!!
                this.pumpCommandType = null
                driverStatusInternal = PumpDriverState.ErrorCommunicatingWithPump
                rxBus.send(EventPumpStatusChanged(driverStatusInternal))
            }
        }
        return null
    }

    fun sleepSeconds(seconds: Long) {
        try {
            Thread.sleep(seconds * 1000)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    fun sleep(miliseconds: Long) {
        try {
            Thread.sleep(miliseconds)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    fun toDateTimeDto(atechDateTimeIn: Long): DateTimeDto {
        var atechDateTime = atechDateTimeIn
        val year = (atechDateTime / 10000000000L).toInt()
        atechDateTime -= year * 10000000000L
        val month = (atechDateTime / 100000000L).toInt()
        atechDateTime -= month * 100000000L
        val dayOfMonth = (atechDateTime / 1000000L).toInt()
        atechDateTime -= dayOfMonth * 1000000L
        val hourOfDay = (atechDateTime / 10000L).toInt()
        atechDateTime -= hourOfDay * 10000L
        val minute = (atechDateTime / 100L).toInt()
        atechDateTime -= minute * 100L
        val second = atechDateTime.toInt()
        return DateTimeDto(year, month, dayOfMonth, hourOfDay, minute, second)
    }

    enum class StatusChange {
        GetStatus, GetCommand, SetStatus, SetCommand, GetError, SetError
    }

    fun getSettingIdAsArray(settingId: Int): ByteArray {
        val array = ByteArray(8)
        CreateGLB_SAFE_VAR(settingId, array)
        return array
    }

    fun isSame(d1: Double, d2: Double): Boolean {
        val diff = d1 - d2
        return Math.abs(diff) <= 0.000001
    }

    fun isSame(d1: Double, d2: Int): Boolean {
        val diff = d1 - d2
        return Math.abs(diff) <= 0.000001
    }

    fun computeUserLevelPassword(btAddress: String): ByteArray {
        val sourceArray = byteArrayOf(91, 215.toByte(), 16, 102, 1, 242.toByte(), 79, 60, 143.toByte(), 107)
        val array = ByteArray(16)
        val address = btAddress.split(":").toTypedArray()
        Log.d("DD", "BTAddress: $address")
        array[0] = convertIntegerStringToByte(address[0])
        array[1] = convertIntegerStringToByte(address[1])
        array[2] = convertIntegerStringToByte(address[2])
        array[3] = convertIntegerStringToByte(address[3])
        array[4] = convertIntegerStringToByte(address[4])
        array[5] = convertIntegerStringToByte(address[5])
        System.arraycopy(sourceArray, 0, array, 6, 10)
        //        return MD5.Create().ComputeHash(array);

//        MessageDigest.getInstance(MD5)
//
//        DigestUtils.md5Hex(password).toUpperCase();
        var md: MessageDigest?
        return try {
            md = MessageDigest.getInstance("MD5")
            md.update(array)
            md.digest()
        } catch (e: NoSuchAlgorithmException) {
            e.printStackTrace()
            array
        }

        //return array;
    }

    private fun convertIntegerStringToByte(hex: String): Byte {
        //Log.d("DD", "BTAddress Part: " + hex);
        return hex.toInt(16).toByte()
    }



    fun getBytesFromInt16(value: Int): ByteArray {
        val array = getBytesFromInt(value)
        Log.d("HHH", ByteUtil.getHex(array))
        return byteArrayOf(array[3], array[2]) // 2, 3
    }

    fun getBytesFromInt(value: Int): ByteArray {
        return ByteBuffer.allocate(4).putInt(value).array()
    }


    fun byteToInt(data: ByteArray, start: Int, length: Int): Int {
        return if (length == 1) {
            ByteUtil.toInt(data[start])
        } else if (length == 2) {
            ByteUtil.toInt(data[start], data[start + 1], ByteUtil.BitConversion.LITTLE_ENDIAN)
        } else if (length == 3) {
            ByteUtil.toInt(data[start], data[start + 1], data[start + 2], ByteUtil.BitConversion.LITTLE_ENDIAN)
        } else if (length == 4) {
            ByteUtil.toInt(data[start], data[start + 1], data[start + 2], data[start + 3], ByteUtil.BitConversion.LITTLE_ENDIAN)
        } else {
            //aapsLogger.error(LTag.PUMPBTCOMM, "byteToInt, length $length not supported.");
            0
        }
    }

    //public static byte[]
    // TOOO fix
    fun CreateGLB_SAFE_VAR(piValueIn: Int, pbyBuffer: ByteArray) {
        var piValue = piValueIn
        pbyBuffer[0] = piValue.toByte()
        piValue = piValue shr 8
        pbyBuffer[1] = piValue.toByte()
        piValue = piValue shr 8
        pbyBuffer[2] = piValue.toByte()
        piValue = piValue shr 8
        pbyBuffer[3] = piValue.toByte()
        pbyBuffer[4] = pbyBuffer[0].inv()
        pbyBuffer[5] = pbyBuffer[1].inv()
        pbyBuffer[6] = pbyBuffer[2].inv()
        pbyBuffer[7] = pbyBuffer[3].inv()
    }

    fun getValueFromeGLB_SAFE_VAR(paBuffer: ByteArray): Int {
        if (paBuffer[7].inv() and 0xFF.toByte() != paBuffer[3] and 0xFF.toByte() ||
            paBuffer[6].inv() and 0xFF.toByte() != paBuffer[2] and 0xFF.toByte() ||
            paBuffer[5].inv() and 0xFF.toByte() != paBuffer[1] and 0xFF.toByte() ||
            paBuffer[4].inv() and 0xFF.toByte() != paBuffer[0] and 0xFF.toByte()) {
            throw InvalidParameterException("Invalid GLB_SAFE_VAR byte array:$paBuffer")
        }
        return byteToInt(paBuffer, 0, 4)
    }

    fun getBytesFromIntArray2(value: Int): ByteArray {
        val array = ByteBuffer.allocate(4).putInt(value).array()
        return byteArrayOf(array[3], array[2])
    }

    fun sendNotification(notificationType: YpsoPumpNotificationType) {
        val notification = Notification( //
            notificationType.notificationType,  //
            resourceHelper.gs(notificationType.resourceId),  //
            notificationType.notificationUrgency)
        rxBus.send(EventNewNotification(notification))
    }

    fun sendNotification(notificationType: YpsoPumpNotificationType, vararg parameters: Any?) {
        val notification = Notification( //
            notificationType.notificationType,  //
            resourceHelper.gs(notificationType.resourceId, *parameters),  //
            notificationType.notificationUrgency)
        rxBus.send(EventNewNotification(notification))
    }

    companion object {

        const val MAX_RETRY = 2
        //private var driverStatus = PumpDriverState.Sleeping

        @JvmStatic private var driverStatusInternal: PumpDriverState = PumpDriverState.Sleeping
        @JvmStatic private var errorTypeInternal: YpsoPumpErrorType? = null

    }
}